### DISCUSSION TITLE
413 Request Entity Too Large Errors

### DISCUSSION THREAD

**Discussion by mjamil** [gotenberg/gotenberg#492] [Q&A]
Hi folks, 

We have a php script that generates word files (stickers/labels to be exact). The script then calls gotenberg to generate a PDF of that word file.

We setup Gotenberg on AWS Elastic Beanstalk using the following dockerrun.aws.json:

```{
    "AWSEBDockerrunVersion":"1",
    "Image":{
        "Name":"gotenberg/gotenberg:7",
        "Update": "true"
    },

    "Ports":[
        {
            "ContainerPort":3000,
            "HostPort":3000
        }
    ]
}
```
This is how we call it from the php script:
```
Gotenberg::libreOffice($apiURL)->outputFilename($output_file)
        ->convert(Stream::path($output_file_path));
```
Everything works great unless we try to convert a word file to pdf that's slightly large.

In that case, we start seeing "413 Request Entity Too Large Errors" in the logs.

Any tips on what to do here? Do we need to increase the CPU or memory on the server? 

Thank you so much!

**Comment by mjamil** [comment]
After we SSHed into the EB server, we were able to edit nginx.conf and set client_max_body_size 300M;

This resolved the error, but now generating the PDF takes an infinite amount of time.

**Reply by gulien** [reply]
Hello @mjamil !

Indeed, Gotenberg does not send `413 Request Entity Too Large` responses 😄 

Regarding the time to convert large documents to PDF, it looks like LibreOffice does struggle to handle such files:

* https://github.com/gotenberg/gotenberg/issues/490
* https://github.com/gotenberg/gotenberg/issues/459
* https://github.com/gotenberg/gotenberg/issues/407
* https://github.com/gotenberg/gotenberg/issues/465

Common "solutions" are:

* Increase the CPU/Memory available.
* Increase the `api-timeout`.

**Comment by mjamil** [comment]
Thanks gulien!

One thing that was not helping at all was long file names with spaces in the file names. Removing that helped a lot. Its still slow, but at least functional. Thank you!

Quick question, do you think switching to unoserver would help?

**Reply by gulien** [reply]
Long file names are causing an issue? That’s odd lol but I’ll investigate to understand why.

`unoserver` Is mostly the same as `unoconv`, and in Gotenbeg I’ve already implemented the « server » part (i.e., starting the LibreOffice instance).

**Reply by mjamil** [reply]
Maybe I was off, It worked once when i made the change to short file names -- so i incorrectly associated the two. Now its back to old behavior  -- which is hang forever waiting for pdf to generate.

**Reply by gulien** [reply]
I wonder if with earlier version of Gotenberg (like `6.4.4`) you’d have the same issue.

**Reply by mjamil** [reply]
Let me give it a shot. Will report back with findings.

**Reply by mjamil** [reply]
Hi @gulien - hope all is well!

Just reporting back. Could not resolve performance issues with 6.x or 7.x when convert word -> pdf (10+pages). 

So we took our precise word template (label files), and converted them to HTML. It took a while but we finally got it working well. Super fast and working great!

Quick question -- any recommendations for what kind of machine we should use to run Gotenberg in production? CPU optimized or memory optimized, or general purpose?

**Reply by gulien** [reply]
Hey @gulien thanks for the feedback! Glad you found a solution 😉

I’d recommend a general purpose instance, not sure you’d gain much with specialized instances.
